fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+            FdMonitorCallback::Ffi(callback, param) => {
+                // Safety: identical objects are generated on both sides by cxx bridge as
+                // integers of the same size (minimum size to fit the enum).
+                let reason = unsafe { std::mem::transmute(ItemWakeReason::Poke) };
+                (callback)(&mut self.fd as *mut _, reason, *param)
+            }
+        }
+        // Return `ItemAction::Remove` if the callback closed the fd
+        match self.fd.is_valid() {
+            true => ItemAction::Retain,
+            false => ItemAction::Remove,
+        }
+    }
+
+    fn new() -> Self {
+        Self {
+            callback: FdMonitorCallback::None,
+            fd: AutoCloseFd2::empty(),
+            timeout: None,
+            last_time: None,
+            item_id: FdMonitorItemId(0),
+        }
+    }
+
+    // fn set_callback_ffi(&mut self, callback: FfiCallback, param: void_ptr) {
+    fn set_callback_ffi(&mut self, callback: *const c_void, param: *const c_void) {
+        // We need to convert between `const * c_void` to a `void_ptr` because we can only implement
+        // traits like `Send` and `Sync` for new types around raw pointers, but cxx bridge doesn't
+        // support passing the opaque type instead `*const c_void` pointer to our POD type.
+        // While we're at, we might as well transmute the correct types instead of using the
+        // equivalent of `void *` to eliminate mistakes on the rust side of things.
+        let callback = unsafe { std::mem::transmute(callback) };
+        let param = unsafe { std::mem::transmute(param) };
+        self.callback = FdMonitorCallback::Ffi(callback, param);
+    }
+}
+
+fn new_fd_monitor_ffi() -> Box<FdMonitor> {
+    FdMonitor::new_ffi()
+}
+
+fn new_fd_monitor_item_ffi(fd: RawFd, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> Box<FdMonitorItem> {
+    // We need to convert between `const * c_void` to a `void_ptr` because we can only implement
+    // traits like `Send` and `Sync` for new types around raw pointers, but cxx bridge doesn't
+    // support passing the opaque type instead `*const c_void` pointer to our POD type.
+    // While we're at, we might as well transmute the correct types instead of using the
+    // equivalent of `void *` to eliminate mistakes on the rust side of things.
+    let callback = unsafe { std::mem::transmute(callback) };
+    let param = unsafe { std::mem::transmute(param) };
+    let mut item = FdMonitorItem::new();
+    item.fd.reset(fd);
+    item.callback = FdMonitorCallback::Ffi(callback, param);
+    if timeout_usecs != FdReadableSet::kNoTimeout {
+        item.timeout = Some(Duration::from_micros(timeout_usecs));
+    }
+    return Box::new(item);
+}
+
+/// A class which can monitor a set of fds, invoking a callback when any becomes readable (or has
+/// been HUP'd) or when per-item-configurable timeouts are reached.
+pub struct FdMonitor {
```

Comment:
Ah I missed that `fd_monitor_t` is being used across threads in the cpp code. That definitely changes how the code should be ported and we'd need to use interior mutability to be able to use `&self` instead of `&mut self` but still make changes. Sorry, I should have caught that!

As for asserting that it is `Sync`, you can actually just hard code such a check as some const code without a test, e.g. https://github.com/neosmart/securestore-rs/blob/54488de686ac5e17833379ad3cb7f90540f4fa74/securestore/src/lib.rs#L200-L210